sandbox/src/run/java.ts

```diff
@@ -30,25 +39,44 @@ export async function runJava(
 		child.stderr.on('data', (data) => {
 			outputBuffer += data.toString();
 		});
+
+		let runStartTime = performance.now();
 		child.stdin.write(input);
 		child.stdin.end();
 
-		let resolved = false;
-
+		let timeLimitExceeded = false;
+		let completedNormally = false;
+		
 		child.on('close', () => {
-			if (!resolved) {
-				resolved = true;
-				resolve(outputBuffer);
+			completedNormally = !timeLimitExceeded;
+
+			let runEndTime = performance.now();
+			const runtimeMilliseconds = Math.floor(runEndTime - runStartTime);
+
+			if (completedNormally) {
+				clearTimeout(timeoutHandle);
+				resolve({kind: RunResultKind.Completed, teamOutput: outputBuffer, 
+					exitCode: child.exitCode!, runtimeMilliseconds});
+			}
+			else {
+				console.log(`Process terminated, total sandbox time: ${runtimeMilliseconds}ms`);
+				resolve({kind: RunResultKind.TimeLimitExceeded, teamOutput: outputBuffer});
 			}
 		});
 
-		setTimeout(() => {
-			if (!resolved) {
-				console.log('30 seconds reached, killing process');
-				resolved = true;
-				child.kill('SIGKILL');
-				resolve(outputBuffer + '\n[Timeout after 30 seconds]');
+		let timeoutHandle = setTimeout(() => {
+			if (completedNormally) {
+				return;
 			}
-		}, 30000);
+
+			console.log(`Run timed out after ${timeoutSeconds} seconds, killing process...`);
+			timeLimitExceeded = true;
+
+			child.stdin.end();
```

Comment:
Do you know if the SIGKILL was actually killing the process for you without the stream end/destroys below? Could maybe be environment related, I don't know a ton about this area.

Reply:
I did remember doing some basic testing to make sure it was killing the process and it seemed to be working.